File: libraries/ndf/ndf_form.f
      SUBROUTINE NDF_FORM( INDF, COMP, FORM, STATUS )
*+
*  Name:
*     NDF_FORM

*  Purpose:
*     Obtain the storage form of an NDF array component.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF_FORM( INDF, COMP, FORM, STATUS )

*  Description:
*     The routine returns the storage form of an NDF array component as
*     an upper case character string (e.g. 'SIMPLE').

*  Arguments:
*     INDF = INTEGER (Given)
*        NDF identifier.
*     COMP = CHARACTER * ( * ) (Given)
*        Name of the NDF array component whose storage form is required:
*        'DATA', 'QUALITY' or 'VARIANCE'.
*     FORM = CHARACTER * ( * ) (Returned)
*        Storage form of the component.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Notes:
*     -  The symbolic constant NDF__SZFRM may be used for declaring the
*     length of a character variable to hold the storage form of an NDF
*     array component. This constant is defined in the include file
*     NDF_PAR.
*     -  At present, the NDF_ routines only support "primitive", "simple",
*     "delta" and "scaled" arrays, so only the values 'PRIMITIVE', 'SIMPLE',
*     'DELTA' and 'SCALED' can be returned.

*  Algorithm:
*     -  Import the NDF identifier.
*     -  Compare the NDF component name with each value in turn,
*     allowing abbreviations.
*     -  Take the appropriate action, or report an error if an
*     inappropriate component name was specified.
*     -  If the component name was not recognised, then report an error.
*     -  If an error occurred, then report context information.

*  Copyright:
*     Copyright (C) 1989, 1990 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     20-OCT-1989 (RFWS):
*        Original version.
*     6-DEC-1989 (RFWS):
*        Added support for the variance component.
*     7-DEC-1989 (RFWS):
*        Changed to give more specific error messages if an
*        inappropriate component name is specified.
*     30-JAN-1990 (RFWS):
*        Installed support for the quality component.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'NDF_PAR'          ! NDF_ public constants
      INCLUDE 'NDF_CONST'        ! NDF_ private constants
      INCLUDE 'NDF_ERR'          ! NDF_ error codes

*  Global Variables:
      INCLUDE 'NDF_ACB'          ! NDF_ Access Control Block
*        ACB_DID( NDF__MXACB ) = INTEGER (Read)
*           ARY_ system identifier for the NDF's data array.

*  Arguments Given:
      INTEGER INDF
      CHARACTER * ( * ) COMP

*  Arguments Returned:
      CHARACTER * ( * ) FORM

*  Status:
      INTEGER STATUS             ! Global status

*  External References:
      LOGICAL NDF1_SIMLR         ! String compare with abbreviation

*  Local variables:
      INTEGER IACB               ! Index to NDF entry in the ACB

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Import the NDF identifier.
      CALL NDF1_IMPID( INDF, IACB, STATUS )
      IF ( STATUS .EQ. SAI__OK ) THEN

*  Compare the component name with each permitted value in turn,
*  allowing abbreviation.

*  AXIS component.
*  ==============
*  Report an error, since this component has no storage form.
         IF ( NDF1_SIMLR( COMP, 'AXIS', NDF__MINAB ) ) THEN
            STATUS = NDF__CNMIN
            CALL ERR_REP( 'NDF_FORM_AXI',
     :      'An AXIS component does not have a storage form ' //
     :      '(possible programming error).', STATUS )

*  DATA component.
*  ==============
*  Obtain the storage form from the ARY_ system data array identifier
*  in the ACB.
         ELSE IF ( NDF1_SIMLR( COMP, 'DATA', NDF__MINAB ) ) THEN
            CALL ARY_FORM( ACB_DID( IACB ), FORM, STATUS )

*  EXTENSION.
*  =========
*  Report an error, since extensions do not have a storage form.
         ELSE IF ( NDF1_SIMLR( COMP, 'EXTENSION', NDF__MINAB ) ) THEN
            STATUS = NDF__CNMIN
            CALL ERR_REP( 'NDF_FORM_EXT',
     :      'An EXTENSION does not have a storage form (possible ' //
     :      'programming error).', STATUS )

*  HISTORY component.
*  =================
*  Report an error, since this component does not have a storage form.
         ELSE IF ( NDF1_SIMLR( COMP, 'HISTORY', NDF__MINAB ) ) THEN
            STATUS = NDF__CNMIN
            CALL ERR_REP( 'NDF_FORM_HIS',
     :      'A HISTORY component does not have a storage form ' //
     :      '(possible programming error).', STATUS )

*  LABEL component.
*  ===============
*  Report an error, since this component does not have a storage form.
         ELSE IF ( NDF1_SIMLR( COMP, 'LABEL', NDF__MINAB ) ) THEN
            STATUS = NDF__CNMIN
            CALL ERR_REP( 'NDF_FORM_LAB',
     :      'A LABEL component does not have a storage form ' //
     :      '(possible programming error).', STATUS )

*  QUALITY component.
*  =================
*  Inspect the quality component to determine its storage form.
         ELSE IF ( NDF1_SIMLR( COMP, 'QUALITY', NDF__MINAB ) ) THEN
            CALL NDF1_QFRM( IACB, FORM, STATUS )

*  TITLE component.
*  ===============
*  Report an error, since this component does not have a storage form.
         ELSE IF ( NDF1_SIMLR( COMP, 'TITLE', NDF__MINAB ) ) THEN
            STATUS = NDF__CNMIN
            CALL ERR_REP( 'NDF_FORM_TIT',
     :      'A TITLE component does not have a storage form ' //
     :      '(possible programming error).', STATUS )

*  UNITS component.
*  ===============
*  Report an error, since this component does not have a storage form.
         ELSE IF ( NDF1_SIMLR( COMP, 'UNITS', NDF__MINAB ) ) THEN
            STATUS = NDF__CNMIN
            CALL ERR_REP( 'NDF_FORM_UNI',
     :      'A UNITS component does not have a storage form ' //
     :      '(possible programming error).', STATUS )

*  VARIANCE component.
*  ==================
*  Inspect the variance component to determine its storage form.
         ELSE IF ( NDF1_SIMLR( COMP, 'VARIANCE', NDF__MINAB ) ) THEN
            CALL NDF1_VFRM( IACB, FORM, STATUS )

*  If the NDF component name was not recognised, then report an error.
         ELSE
            STATUS = NDF__CNMIN
            CALL MSG_SETC( 'BADCOMP', COMP )
            CALL ERR_REP( 'NDF_FORM_COMP',
     :                    'Invalid array component name ' //
     :                    '''^BADCOMP'' specified (possible ' //
     :                    'programming error).', STATUS )
         END IF
      END IF

*  If an error occurred, then report context information and call the
*  error tracing routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         CALL ERR_REP( 'NDF_FORM_ERR',
     :   'NDF_FORM: Error obtaining the storage form of an NDF ' //
     :   'array component.', STATUS )
         CALL NDF1_TRACE( 'NDF_FORM', STATUS )
      END IF

      END

File: libraries/ndf/ndfForm.c
#include <stdlib.h>
#include "sae_par.h"
#include "dat_par.h"
#include "ndf1.h"
#include "ndf_err.h"
#include "ndf.h"
#include "mers.h"

void ndfForm_( int indf, const char *comp, char *form, size_t form_length,
              int *status ){
/*
*+
*  Name:
*     ndfForm

*  Purpose:
*     Obtain the storage form of an NDF array component.

*  Synopsis:
*     void ndfForm( int indf, const char *comp, char *form,
*                   size_t form_length, int *status )

*  Description:
*     This function returns the storage form of an NDF array component as
*     an upper case character string (e.g. "SIMPLE").

*  Parameters:
*     indf
*        NDF identifier.
*     comp
*        Pointer to a null terminated string holding the name of the NDF
*        array component whose storage form is required: "DATA", "QUALITY"
*        or "VARIANCE".
*     form
*        Pointer to an array in which to return a null terminated string
*        holding the storage form of the component.
*     form_length
*        The length of the supplied 'form' array. This should include
*        room for the terminating null.
*     *status
*        The global status.

*  Notes:
*     -  The symbolic constant NDF__SZFRM may be used for declaring the
*     length of a character variable to hold the storage form of an NDF
*     array component. This constant is defined in the header file "ndf.h".
*     -  At present, the NDF_ functions only support "primitive", "simple",
*     "delta" and "scaled" arrays, so only the values "PRIMITIVE",
*     "SIMPLE", "DELTA" and "SCALED" can be returned.

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     xxx (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local variables: */
   NdfACB *acb;          /* Pointer to NDF entry in the ACB */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Ensure the NDF library has been initialised. */
   NDF_INIT( status );

/* Import the NDF identifier. */
   ndf1Impid( indf, &acb, status );
   if( *status == SAI__OK ) {

/* Compare the component name with each permitted value in turn,
   allowing abbreviation. */

/* AXIS component.
   ==============
   Report an error, since this component has no storage form. */
      if( ndf1Simlr( comp, 1, 0, "AXIS", NDF__MINAB ) ) {
         *status = NDF__CNMIN;
         errRep( " ", "An AXIS component does not have a storage form "
                 "(possible programming error).", status );

/* DATA component.
   ==============
   Obtain the storage form from the ARY_ system data array identifier
   in the ACB. */
      } else if( ndf1Simlr( comp, 1, 0, "DATA", NDF__MINAB ) ) {
         aryForm( acb->did, form, status );

/* EXTENSION.
   =========
   Report an error, since extensions do not have a storage form. */
      } else if( ndf1Simlr( comp, 1, 0, "EXTENSION", NDF__MINAB ) ) {
         *status = NDF__CNMIN;
         errRep( " ", "An EXTENSION does not have a storage form (possible "
                 "programming error).", status );

/* HISTORY component.
   =================
   Report an error, since this component does not have a storage form. */
      } else if( ndf1Simlr( comp, 1, 0, "HISTORY", NDF__MINAB ) ) {
         *status = NDF__CNMIN;
         errRep( " ", "A HISTORY component does not have a storage form "
                 "(possible programming error).", status );

/* LABEL component.
   ===============
   Report an error, since this component does not have a storage form. */
      } else if( ndf1Simlr( comp, 1, 0, "LABEL", NDF__MINAB ) ) {
         *status = NDF__CNMIN;
         errRep( " ", "A LABEL component does not have a storage form "
                 "(possible programming error).", status );

/* QUALITY component.
   =================
   Inspect the quality component to determine its storage form. */
      } else if( ndf1Simlr( comp, 1, 0, "QUALITY", NDF__MINAB ) ) {
         ndf1Qfrm( acb, form, form_length, status );

/* TITLE component.
   ===============
   Report an error, since this component does not have a storage form. */
      } else if( ndf1Simlr( comp, 1, 0, "TITLE", NDF__MINAB ) ) {
         *status = NDF__CNMIN;
         errRep( " ", "A TITLE component does not have a storage form "
                 "(possible programming error).", status );

/* UNITS component.
   ===============
   Report an error, since this component does not have a storage form. */
      } else if( ndf1Simlr( comp, 1, 0, "UNITS", NDF__MINAB ) ) {
         *status = NDF__CNMIN;
         errRep( " ", "A UNITS component does not have a storage form "
                 "(possible programming error).", status );

/* VARIANCE component.
   ==================
   Inspect the variance component to determine its storage form. */
      } else if( ndf1Simlr( comp, 1, 0, "VARIANCE", NDF__MINAB ) ) {
         ndf1Vfrm( acb, form, form_length, status );

/* If the NDF component name was not recognised, then report an error. */
      } else {
         *status = NDF__CNMIN;
         msgSetc( "BADCOMP", comp );
         errRep( " ", "Invalid array component name '^BADCOMP' specified "
                 "(possible programming error).", status );
      }
   }

/* If an error occurred, then report context information and call the
   error tracing function. */
   if( *status != SAI__OK ) {
      errRep( " ", "ndfForm: Error obtaining the storage form of an NDF "
              "array component.", status );
      ndf1Trace( "ndfForm", status );
   }

/* Restablish the original AST status pointer */
   NDF_FINAL

}

